### ISSUE TITLE
memory mode map initialization slow when database is too big

### ISSUE THREAD

**Issue by ktsaou** [netdata/netdata#2382]
When the database is too big (several GBs) and `memory mode = map` is used, netdata initializes the whole memory with zeros, which results in all these GBs to be flushed to disk. We should be able to detect these are just initialized data and avoid this duplicate initialization (so the files on disk will not be re-written - they will remain sparse files).

Similarly, when a slave host is offline and reconnects, netdata either initializes all its database to zero, or it fills in a gap (depending on the duration it was offline compared to the duration of the database). Depending on the duration the slave was offline, this again leads to huge amounts of data being flushed to disk.

**Comment by Ferroin** [comment]
For the initialization side of things, had you considered using fallocate() prior to mapping the file?  That would eliminate the need to zero-fill the memory (dependent on what flags you use), and on filesystems that support unwritten extents (almost everything used on Linux these days, as well as ZFS and I think HFS+ too), it would be far faster, while also guaranteeing (unless you're using BTRFS) that you'll be able to write out all the data.

**Comment by Ferroin** [comment]
As a more specific performance example, fallocate() creating a 1G file on BTRFS on an SSD takes about 0.008s on my system, while doing the same by writing out null bytes in 4k chunks takes roughly 1.1s.

**Comment by ktsaou** [comment]
The way it is now, a sparse file is created (just a byte is written at the wanted file size - almost zero time) and then it is `mmap()`'d.

But then netdata detects that the memory mapped from that file, does not contain something it can use (old data), and fills the entire memory block with zeros with `memset()`, which results in huge amounts of data being flushed to disk.

This optimization is about passing back a flag that states that this was a new file, to prevent the call to `memset()`.

The same `memset()` is made if the data loaded do not overlap with the current time-range of the database (ie. if the database is 1 hour long, the last point in the file is older than 1 hour).

**Comment by Ferroin** [comment]
Ah, that makes sense.  It may still be worth having the option to fallocate instead though, since if that data isn't written out, you have no guarantee that any arbitrarily written values until the database history limit won't have errors when trying to write-back (the initial zeroing of the file also makes sure the space is allocated for it).  I don't do enough low-level stuff in C to know how exactly ENOSPC is handled in a write to an mmaped file though, so I may just be worrying about nothing.

**Comment by GitStoph** [comment]
![image](https://user-images.githubusercontent.com/20820226/27540285-4702f24a-5a45-11e7-9578-dc8ba1f49541.png)
I wasn't sure if here, or if #2364 was the proper place for this. This was the result of a simple `sudo systemctl restart netdata` after commenting out settings for a backend server. 
```
# enabled = yes
# update every = 3
```
However, re-running the netdata updater script appeared to have settled it down.. After doing a little low level testing, I may have found a pattern. Running `sudo systemctl restart netdata` when there are no changes to the master's `netdata.conf` file has consistently yielded no issues.  However, any time I make a change to the `netdata.conf` file, however minimal, the restart will either crash the process, or clog the cpu with what is likely upstream server data? 
![image](https://user-images.githubusercontent.com/20820226/27541072-2f27fbe0-5a48-11e7-94b8-75e34b2d4a39.png)

This was the result of uncommenting `update ever = 1` under `[global]`.  Here are some potentially useful journal logs:
```
-- The start-up result is done.
Jun 26 08:15:32 netdata sudo[29375]: pam_unix(sudo:session): session closed for user root
Jun 26 08:15:33 netdata kernel: netdata[29640]: segfault at 1260 ip 0000000000437aae sp 00007
Jun 26 08:15:36 netdata systemd[1]: netdata.service: Main process exited, code=killed, status
Jun 26 08:15:36 netdata systemd[1]: netdata.service: Unit entered failed state.
Jun 26 08:15:36 netdata systemd[1]: netdata.service: Failed with result 'signal'.
-- Unit netdata.service has begun shutting down.
Jun 26 08:22:20 netdata kernel: netdata invoked oom-killer: gfp_mask=0x24280ca, order=0, oom_
Jun 26 08:22:20 netdata kernel: netdata cpuset=/ mems_allowed=0
Jun 26 08:22:20 netdata kernel: CPU: 0 PID: 30412 Comm: netdata Not tainted 4.4.0-21-generic
Jun 26 08:22:20 netdata kernel: Hardware name: VMware, Inc. VMware Virtual Platform/440BX Des
Jun 26 08:22:20 netdata kernel:  0000000000000286 0000000087b913aa ffff880100187b28 ffffffff8
Jun 26 08:22:20 netdata kernel:  ffff880100187d20 ffff8801ee6d8dc0 ffff880100187b98 ffffffff8
Jun 26 08:22:20 netdata kernel:  0000000000000015 ffffffff81e42860 ffff8800355a5b40 fffffffff
Jun 26 08:22:20 netdata kernel: Call Trace:
Jun 26 08:22:20 netdata kernel:  [<ffffffff813e93c3>] dump_stack+0x63/0x90
Jun 26 08:22:20 netdata kernel:  [<ffffffff81208b2e>] dump_header+0x5a/0x1c5
Jun 26 08:22:20 netdata kernel:  [<ffffffff8133a0cd>] ? cap_capable+0xd/0x70
Jun 26 08:22:20 netdata kernel:  [<ffffffff81190dd2>] oom_kill_process+0x202/0x3c0
Jun 26 08:22:20 netdata kernel:  [<ffffffff811911f5>] out_of_memory+0x215/0x460
Jun 26 08:22:20 netdata kernel:  [<ffffffff81197190>] __alloc_pages_nodemask+0x9d0/0xb60
Jun 26 08:22:20 netdata kernel:  [<ffffffff810ba1bc>] ? set_next_entity+0x9c/0xb0
Jun 26 08:22:20 netdata kernel:  [<ffffffff811e209e>] alloc_pages_vma+0xbe/0x240
Jun 26 08:22:20 netdata kernel:  [<ffffffff8181fd36>] ? __schedule+0x386/0xa10
Jun 26 08:22:20 netdata kernel:  [<ffffffff811c005e>] handle_mm_fault+0x148e/0x1820
Jun 26 08:22:20 netdata kernel:  [<ffffffff810abd65>] ? wake_up_process+0x15/0x20
Jun 26 08:22:20 netdata kernel:  [<ffffffff813f7834>] ? call_rwsem_down_read_failed+0x14/0x30
Jun 26 08:22:20 netdata kernel:  [<ffffffff8106b537>] __do_page_fault+0x197/0x400
Jun 26 08:22:20 netdata kernel:  [<ffffffff8106b7c2>] do_page_fault+0x22/0x30
Jun 26 08:22:20 netdata kernel:  [<ffffffff81826678>] page_fault+0x28/0x30
Jun 26 08:22:20 netdata kernel: Mem-Info:
Jun 26 08:22:20 netdata kernel: active_anon:1645537 inactive_anon:324459 isolated_anon:0
                                 active_file:84 inactive_file:77 isolated_file:0
                                 unevictable:915 dirty:0 writeback:119 unstable:0
                                 slab_reclaimable:16806 slab_unreclaimable:10779
                                 mapped:730 shmem:139 pagetables:11026 bounce:0
                                 free:12608 free_pcp:1050 free_cma:0
Jun 26 08:22:20 netdata kernel: Node 0 DMA free:15868kB min:20kB low:24kB high:28kB active_an
Jun 26 08:22:20 netdata kernel: lowmem_reserve[]: 0 2973 7964 7964 7964
Jun 26 08:22:20 netdata kernel: Node 0 DMA32 free:25256kB min:4232kB low:5288kB high:6348kB a
Jun 26 08:22:20 netdata kernel: lowmem_reserve[]: 0 0 4990 4990 4990
Jun 26 08:22:20 netdata kernel: Node 0 Normal free:9308kB min:7104kB low:8880kB high:10656kB
Jun 26 08:22:20 netdata kernel: lowmem_reserve[]: 0 0 0 0 0
Jun 26 08:22:20 netdata kernel: Node 0 DMA: 1*4kB (U) 1*8kB (U) 1*16kB (U) 3*32kB (U) 2*64kB
Jun 26 08:22:20 netdata kernel: Node 0 DMA32: 196*4kB (UME) 173*8kB (UME) 123*16kB (UME) 102*
Jun 26 08:22:20 netdata kernel: Node 0 Normal: 368*4kB (UME) 242*8kB (UME) 131*16kB (UME) 63*
Jun 26 08:22:20 netdata kernel: Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hu
Jun 26 08:22:20 netdata kernel: Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hu
Jun 26 08:22:20 netdata kernel: 1837 total pagecache pages
Jun 26 08:22:20 netdata kernel: 898 pages in swap cache
Jun 26 08:22:20 netdata kernel: Swap cache stats: add 4637613, delete 4636715, find 67569/102
Jun 26 08:22:20 netdata kernel: Free swap  = 0kB
Jun 26 08:22:20 netdata kernel: Total swap = 8384508kB
Jun 26 08:22:20 netdata kernel: 2097038 pages RAM
Jun 26 08:22:20 netdata kernel: 0 pages HighMem/MovableOnly
Jun 26 08:22:20 netdata kernel: 53227 pages reserved
Jun 26 08:22:20 netdata kernel: 0 pages cma reserved
Jun 26 08:22:20 netdata kernel: 0 pages hwpoisoned
Jun 26 08:22:20 netdata kernel: [ pid ]   uid  tgid total_vm      rss nr_ptes nr_pmds swapent
Jun 26 08:22:20 netdata kernel: [  521]     0   521    11164      388      24       3      79
Jun 26 08:22:20 netdata kernel: [  576]     0   576    44176      131      20       3      59
Jun 26 08:22:20 netdata kernel: [  583]     0   583    11167      307      25       3      29
Jun 26 08:22:20 netdata kernel: [  975]   100   975    25082      397      19       3       5
Jun 26 08:22:20 netdata kernel: [ 1097]     0  1097     7253      379      18       3       7
Jun 26 08:22:20 netdata kernel: [ 1103]     0  1103     6512      362      17       3       5
Jun 26 08:22:20 netdata kernel: [ 1106]   104  1106    64101      207      27       3      21
Jun 26 08:22:20 netdata kernel: [ 1111]   107  1111    10726      418      25       3       8
Jun 26 08:22:20 netdata kernel: [ 1143]     0  1143   187963        0      32       3      79
Jun 26 08:22:20 netdata kernel: [ 1145]     0  1145    28284       52      47       3      56
Jun 26 08:22:20 netdata kernel: [ 1150]     0  1150     7163      238      19       3       7
Jun 26 08:22:20 netdata kernel: [ 1156]     0  1156    68965      222      38       4      17
Jun 26 08:22:20 netdata kernel: [ 1158]     0  1158   109089      638      36       6     278
Jun 26 08:22:20 netdata kernel: [ 1166]     0  1166     1101      286       8       3       3
Jun 26 08:22:20 netdata kernel: [ 1187]     0  1187     3345        1      11       3       3
Jun 26 08:22:20 netdata kernel: [ 1192]     0  1192    69273        2      40       4      16
Jun 26 08:22:20 netdata kernel: [ 1284]     0  1284    16381      407      36       3      18
Jun 26 08:22:20 netdata kernel: [ 1292]     0  1292     1307        4       8       3       2
Jun 26 08:22:20 netdata kernel: [ 1293]     0  1293     1432      880       8       3
Jun 26 08:22:20 netdata kernel: [ 1344]     0  1344     3986      316      12       3       4
Jun 26 08:22:20 netdata kernel: [ 1351]     0  1351     4907      377      13       3       4
Jun 26 08:22:20 netdata kernel: [13193]   108 13193     7040        7      17       3       5
Jun 26 08:22:20 netdata kernel: [19555]     0 19555    23853      439      49       3      23
Jun 26 08:22:20 netdata kernel: [19573]  1000 19573    11313      360      25       3      22
Jun 26 08:22:20 netdata kernel: [19575]  1000 19575    15337        1      32       3      51
Jun 26 08:22:20 netdata kernel: [19655]  1000 19655    23853      266      47       3      25
Jun 26 08:22:20 netdata kernel: [19660]  1000 19660     5655      415      17       3      48
Jun 26 08:22:20 netdata kernel: [29662]   999 29662  5717435  1965560   10253      30  208486
Jun 26 08:22:20 netdata kernel: [30794]     0 30794    13177      391      30       3      11
Jun 26 08:22:20 netdata kernel: [30795]     0 30795     3148      368      10       3       7
Jun 26 08:22:20 netdata kernel: [30814]     0 30814     3373      409      12       3      27
Jun 26 08:22:20 netdata kernel: [ 2929]     0  2929     5815       38      15       3       5
Jun 26 08:22:20 netdata kernel: [ 2931]     0  2931     3853      136      12       4       3
Jun 26 08:22:20 netdata kernel: Out of memory: Kill process 29662 (netdata) score 1980 or sac
Jun 26 08:22:20 netdata kernel: Killed process 29662 (netdata) total-vm:22869740kB, anon-rss:
Jun 26 08:22:21 netdata systemd[1]: netdata.service: Main process exited, code=killed, status
Jun 26 08:22:21 netdata systemd[1]: Stopped Real time performance monitoring.
```
I recommented the `netdata.conf` file back to `# update ever = 1` and ran `sudo systemctl restart netdata`. This is what produced as a result:
```
-- Unit netdata.service has begun shutting down.
Jun 26 08:29:57 netdata /usr/lib/snapd/snapd[1158]: snapmgr.go:422: No snaps to auto-refresh
Jun 26 08:29:57 netdata snapd[1158]: 2017/06/26 08:29:57.564059 snapmgr.go:422: No snaps to a
Jun 26 08:30:26 netdata systemd[1]: netdata.service: State 'stop-sigterm' timed out. Killing.
Jun 26 08:30:26 netdata systemd[1]: netdata.service: Main process exited, code=killed, status
Jun 26 08:30:26 netdata systemd[1]: Stopped Real time performance monitoring.
-- Subject: Unit netdata.service has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit netdata.service has finished shutting down.
Jun 26 08:30:26 netdata systemd[1]: netdata.service: Unit entered failed state.
Jun 26 08:30:26 netdata systemd[1]: netdata.service: Failed with result 'signal'.
Jun 26 08:30:26 netdata systemd[1]: Started Real time performance monitoring.
```
I'm not sure if this is useful to you or not.. stability has returned completely afterwards with no hints of CPU issues or RAM flooding.

**Comment by ktsaou** [comment]
@GitStoph what change did you make to netdata.conf

**Comment by ktsaou** [comment]
@GitStoph to trace the issue I need your input. If you can identify the change you made, I may be able to reproduce the problem so that I can fix it.

**Comment by GitStoph** [comment]
Sorry for the late reply.

While doing some log hunting, I realized that I had made a mistake and had the following netdata.conf
```
[backend]
	# host tags = 
	 enabled = yes
	# data source = average
	# type = graphite
	# destination = localhost
	# prefix = netdata
	# hostname = NetdataMaster
	 update every = 3
	# buffer on failures = 10
	# timeout ms = 20000
	# send charts matching = *
```
Somewhere between interruptions, I had changed my plan and opted to use prometheus instead of graphite. After I realized that netdata was showing backend errors in the log (since there was no localhost graphite datasource existing or listening), I re-commented the config to show the following:
```
[backend]
	# host tags = 
	# enabled = no
	# data source = average
	# type = graphite
	# destination = localhost
	# prefix = netdata
	# hostname = NetdataMaster
	# update every = 10
	# buffer on failures = 10
	# timeout ms = 20000
	# send charts matching = *
```
Essentially, I just restored the `netdata.conf` back to default settings for that section. No other changes were made. At that point I ran `sudo systemctl restart netdata` and that's when the issue occurred. 

I wish I could recall more details. I should have been paying more attention to github mentions. Apologies.

**Comment by ktsaou** [comment]
ok. If it happens again, just open a github issue. I hate bugs, especially those that crash netdata. So, I need your help to find if there are any.

**Comment by GitStoph** [comment]
Hi @ktsaou . This is occurring again. Seems like when I update netdata on OSX clients, they begin to send a duplicate stream which I can see in the error logs on my master.
```
2017-09-26 09:26:22: netdata INFO : STREAM Server4 [receive from [10.2.1.13]:53493]: multiple streaming connections for the same host detected. If multiple netdata are pushing metrics for the same charts, at the same time, the result is unexpected.
2017-09-26 09:31:28: netdata INFO : STREAM Share1 [receive from [10.1.4.40]:50607]: multiple streaming connections for the same host detected. If multiple netdata are pushing metrics for the same charts, at the same time, the result is unexpected.
2017-09-26 09:33:24: netdata INFO : STREAM Share1 [receive from [10.1.4.40]:50619]: multiple streaming connections for the same host detected. If multiple netdata are pushing metrics for the same charts, at the same time, the result is unexpected.
2017-09-26 09:35:00: netdata INFO : STREAM Share1 [receive from [10.1.4.40]:50649]: multiple streaming connections for the same host detected. If multiple netdata are pushing metrics for the same charts, at the same time, the result is unexpected.
2017-09-26 09:35:00: netdata INFO : STREAM Server4 [receive from [10.2.1.13]:53670]: multiple streaming connections for the same host detected. If multiple netdata are pushing metrics for the same charts, at the same time, the result is unexpected.
```

The fix is the same. Stopping netdata on my master, clearing out the databases from `/var/lib/netdata/`, `sudo killall netdata` on the offending OSX machines, then `sudo systemctl start netdata` on the master again. At that point normal functionality resumes. Leaving it overnight does not ever resume normal operation.

**Comment by ktsaou** [comment]
Could you please grep all `Share1` entries from the log? I would like to know how is it possible for a client to connect multiple times in parallel.

I think I may have to exclusively lock each host on only one sender at the master. But it would great to know what sequence of events the master encountered.

**Comment by GitStoph** [comment]
https://pastebin.com/xeqvCtQ7
Here's the pastebin. I sanitized the IP, GUID, and such. Let me know if there's anything else I can grab!

**Comment by GitStoph** [comment]
Sorry for the delay, work has been busy. On the sending netdata, should I specifically grep for the "stream" lines to narrow down the information? Or do you want it all in totality? I believe at the moment i was running the latest available netdata, since the update process is what seemed to cause the issue.

**Comment by GitStoph** [comment]
I was able to recreate the duplicate stream, but you must have added something to the code to prevent duplicate streams on the master? 
```
2017-09-29 10:34:37: netdata INFO : STREAM Share1 [receive from [share1.ip.address]:56402]: multiple streaming connections for the same host detected. If multiple netdata are pushing metrics for the same charts, at the same time, the result is unexpected.
```

It looks like updating with the netdata-updater.sh that comes in the git clone fail's to kill any previous running instances of netdata. Probably because it needs sudo access, or to be run as the netdata user if people chose to create one? 

I was able to curb the behavior by writing a quick and dirty script that essentially kills netdata, waits 5 seconds for it to stop/close every single process, then runs the updater. 3 seconds or less seemed to be too short of a wait, occasionally causing duplicate streams/netdata instances on the machine. For reference: (extra lines in there for logging verbosity)

```
#!/bin/bash
###############################
# Netdata Updater Script 
# 9.29.17 Chris Meisch
###############################

#Set Variables
touch /var/log/netdataupdater.log 
log=/var/log/netdataupdater.log 
uptime=$(last reboot | head -1 | cut -c 37-53) 
host=$(hostname | awk '{print$1}') 
curUser=`python -c 'from SystemConfiguration import SCDynamicStoreCopyConsoleUser; import sys; username = (SCDynamicStoreCopyConsoleUser(None, None, None) or [None])[0]; username = [username,""][username in [u"loginwindow", None, u""]]; sys.stdout.write(username + "\n");'`
NOW="$(date +"%Y-%m-%d%H:%M")" /bin/echo "Netdata Updater started $NOW"

#Start of Script
/bin/echo "Script starting $NOW" >> $log
/bin/echo "$host" >> $log
/bin/echo "Server has been running since $uptime" >> $log
/bin/echo "Log is located at $log" >> $log
/bin/echo "Script is running as $curUser" >> $log

# Running the actual update
/bin/echo "$NOW Killing netdata" >> $log
sudo killall netdata 
sleep 5
/bin/echo "$NOW Updating netdata" >> $log
/Users/<mylocaladmin>/netdata/netdata-updater.sh -f
sleep 5

# Chowning Netdata back to <mylocaladmin>
/bin/echo "$NOW chowning netdata back to <mylocaladmin>." >> $log
sudo chown -R <mylocaladmin>:<mylocaladminsgroup> /Users/<mylocaladmin>/opt/netdata/usr/share/netdata/*

# Restarting Netdata
/bin/echo "$NOW restarting netdata as root" >> $log
sudo /Users/<mylocaladmin>/opt/netdata/usr/sbin/netdata

# Done
/bin/echo "Update has completed $NOW" >> $log

exit 0
```

**Comment by ktsaou** [comment]
So, you say that the slaves were running multiple netdata?

> I was able to recreate the duplicate stream, but you must have added something to the code to prevent duplicate streams on the master?

no, there is nothing to prevent this. I think I should add something though.

> It looks like updating with the netdata-updater.sh that comes in the git clone fail's to kill any previous running instances of netdata. Probably because it needs sudo access, or to be run as the netdata user if people chose to create one?

Generally, the updater does not kill netdata (it only sends it a signal to save its db, if any).
The installer (that is run by the updater) restarts netdata.

Do you have the output of such a run? It would be very helpful if you could help me figure out how to fix that.

**Comment by GitStoph** [comment]
> So, you say that the slaves were running multiple netdata?

Yes, I believe so. An htop filter showed two different instances and different PIDs, and the stream log on the master also indicated the same. 

>no, there is nothing to prevent this. I think I should add something though.
It looks like updating with the netdata-updater.sh that comes in the git clone fail's to kill any previous running instances of netdata. Probably because it needs sudo access, or to be run as the netdata user if people chose to create one?

This is not so? I just assumed there was I guess from this line: `[root /Users/wssadmin/netdata]# stop_all_netdata  ... OK`

>Do you have the output of such a run? It would be very helpful if you could help me figure out how to fix that.

Would a netdata-installer.log suffice?
[netdata-installer.log](https://github.com/firehol/netdata/files/1349731/netdata-installer.log)

**Comment by ktsaou** [comment]
> This is not so? I just assumed there was I guess from this line: [root /Users/wssadmin/netdata]# stop_all_netdata ... OK

ah! ok, this is by the installer (which is run by the updater). And yes, it stops all netdata instances, except the ones running in containers (it checks their namespaces).

So, as I understand it there is no bug in netdata. Right?

**Comment by GitStoph** [comment]
If this is by design, then there is no bug. Just a nuance to be noted by OSX users that netdata needs to be killed first, wait 5 seconds, then run the updater. Otherwise they'll have duplicate instances running on that client which can cause the master to blow up CPU. Although with recent versions of netdata, I haven't had teh CPU issue reoccur?

**Comment by ktsaou** [comment]
> If this is by design, then there is no bug.

Well, ideally we should lock the host each time a slave pushes metrics to it. The problem is that under certain network conditions, the master may believe a slave is connected, although it is not. In that case, if we had locked the host and new socket will be rejected, losing metrics. I think the right solution would be new connections to break the old ones. So, the host will be locked, and when a new streaming request comes for a host, the master will stop the old one for this host.

> Just a nuance to be noted by OSX users that netdata needs to be killed first, wait 5 seconds, then run the updater. Otherwise they'll have duplicate instances running on that client which can cause the master to blow up CPU. Although with recent versions of netdata, I haven't had teh CPU issue reoccur?

I think this has been fixed.

**Comment by GitStoph** [comment]
The last two days I've been seeing this issue again on the master. It always starts immediately after the master runs the cron daily jobs, which is running the netdata updater. Every time it comes back, the CPU goes nuts. 

Was digging through the logs, saw that I was hitting the Ubuntu Server default open file limit on about 21 servers slaved. I'll try increasing the open file limit to see if it resolves the issue. Mostly an FYI.

**Comment by ktsaou** [comment]
hm... did your server run out of file descriptors or just netdata?

Generally, netdata should make a pretty small use of file descriptors. On a master with 30+ slaves, it uses less than 100 descriptors. You can see all the file descriptors netdata has, with `ls -l /proc/NETDATA_PID/fd/`. You should see the data collection files of the local netdata, a socket for each slave and a health-log file for each slave. So, just 2 sockets per slave + local data collection.

If Linux run out of file descriptors, then things could go crazy.

Are you running the latest netdata? I have made even more changes to ensure slaves will not bombard the server with connections in case something is wrong. So, please update once more.

**Comment by GitStoph** [comment]
Not completely sure. After increasing the limit for the netdata user/system, i didn't have that issue in the logs anymore. The issue reoccurred a couple times that day when i tested it. I ended up deleting everything netdata related on the master and reinstalled netdata from scratch again to also ensure I had the latest version of everything. It seems to be behaving again. I absent-mindedly cleared the logs as a result of this deleting however.. Very sorry about that. But on the plus side, with the latest code it again seems to be stable?!

**Comment by ktsaou** [comment]
ok. Have you updated the slaves too? I made a few changes to the slaves' code to ensure they will never bombard the server with connections.

**Comment by GitStoph** [comment]
They were set to run the daily updater, but I'll go back through and check.
`2017-10-13 07:45:43: netdata ERROR: 3466822: Failed to accept new incoming connection. (errno 24, Too many open files)` has come back. Unsure where the bottleneck is currently? 
```
~$ uname -a
Linux netdata 4.4.0-96-generic #119-Ubuntu SMP Tue Sep 12 14:59:54 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
I ran `sudo systemctl stop netdata`, followed by `sudo rm -R /var/lib/netdata/*` and `sudo rm -R /var/cache/netdata/*` . I then ran `sudo systemctl start netdata` and although the slave's db's show in the registry, i cant navigate to them. The browser shows `Cannot open file: /usr/share/netdata/web/index.html` although nothing was done that should have changed permissions. CPU has been running hot at about 50% afterwards. 
I ran `cat /var/log/netdata/error.log | grep ERROR` and got the following: https://pastebin.com/VkQD4uV5

I ran `sudo systemctl restart netdata` one more time and grep'd the logs again. They were again littered with `2017-10-13 09:21:03: netdata ERROR: 60740780: Failed to accept new incoming connection. (errno 24, Too many open files)`

```
~$ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 31775
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 200000
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 31775
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```
Any other info I could provide that may help?

**Comment by ktsaou** [comment]
@GitStoph you say that you restarted netdata and then you get:

```
2017-10-13 09:21:03: netdata ERROR: 60740780: Failed to accept new incoming connection. (errno 24, Too many open files)
```

60.740.780 is a connections counter. When netdata starts, it is set to 0  and is incremented by one every time someone connects to port 19999. So, 60 million connections are really a lot. This netdata is bombarded with connections.

How many slaves are streaming metrics?
Are you sure you have not created a streaming loop? 

---

#### Regarding the slow start

When you `rm -R /var/cache/netdata/*`, this netdata will need to re-create a lot of files when it starts. Depending on the number of the slaves, this may take some time.

Normally you should not need to `rm -R /var/lib/netdata/*`. This dir has only state information.

---

#### Regarding the open files limit

When netdata runs out of file descriptors, could you please post the output of this:

```sh
ls -l /proc/$(pidof netdata)/fd/
```

**Comment by GitStoph** [comment]
I'm not sure how it could have created a streaming loop? As far as netdata goes, there are only 23 servers currently with nothing fancy happening network-wise. They're all set to stream only to the one master, no proxies or anything like that. They all use the same API key which may not be the best practice? 

Three servers are behind a load balancer, but the VIP for that is set via DNS. By streaming directly by their IP, it should be bypassing the load balancer. There's a prometheus server that's scraping the netdata that is running netdata itself.. that'd be the closest to redundant connections between any of the servers?

**Comment by GitStoph** [comment]
May have found a potential loop creator. At some point with one of the OSX netdata updates, it added an update line to `/etc/periodic/daily/netdata-updater`. Previously that wasn't working and I still had a cronjob running the update script I had written. This could have been creating multiple instances on the OSX servers. Currently netdata doesn't want to run on it, so I'll take a look at what may be causing that issue.

**Comment by ktsaou** [comment]
ok, I am sure there is something there. 60 mil connections is not easy to accomplish without a loop...

**Comment by ktsaou** [comment]
This will be fixed with the new database format.
So I am closing this in favor of #4687
